torch/csrc/autograd/profiler.cpp

```diff
@@ -100,43 +135,50 @@ void pushRange(
       cuda_stubs->nvtxRangePushA(name.str());
     }
   } else {
-    getEventList().record(
+    auto& list = getEventList();
+    std::lock_guard<std::mutex> guard(all_event_lists_map_mutex);
+    list.record(
         EventKind::PushRange,
         name,
         thread_id,
-        state == ProfilerState::CUDA,
+        config.state == ProfilerState::CUDA,
         std::move(shapes));
   }
 }
 
 void popRange() {
-  if (state == ProfilerState::Disabled) {
+  if (config.state == ProfilerState::Disabled) {
     return;
   }
-  if (state == ProfilerState::NVTX) {
+  if (config.state == ProfilerState::NVTX) {
     cuda_stubs->nvtxRangePop();
   } else {
-    getEventList().record(
+    auto& list = getEventList();
+    std::lock_guard<std::mutex> guard(all_event_lists_map_mutex);
+    list.record(
         EventKind::PopRange,
         StringView(""),
         thread_id,
-        state == ProfilerState::CUDA);
+        config.state == ProfilerState::CUDA);
   }
 }
 
-void enableProfiler(ProfilerConfig config) {
-  ProfilerState new_state = config.state;
-  AT_ASSERT(new_state != ProfilerState::Disabled);
-  if (new_state == ProfilerState::NVTX && !cuda_stubs->enabled())
-    throw std::runtime_error("Can't use NVTX profiler - PyTorch was compiled without CUDA");
-  if (state != ProfilerState::Disabled && new_state != state) {
-    throw std::runtime_error("can't change kind of profiling (e.g. NVTX to CPU) while profiler is running");
+void enableProfiler(ProfilerConfig new_config, bool emit_start) {
+  TORCH_CHECK(new_config.state != ProfilerState::Disabled);
+  if (config.state != ProfilerState::Disabled) {
+    TORCH_WARN("Trying to enable profiler when profiler is already active,"
+        " new profiler is inactive");
+    ++recursion_depth_;
+    return;
   }
 
+  TORCH_CHECK(new_config.state != ProfilerState::NVTX || cuda_stubs->enabled(),
+      "Can't use NVTX profiler - PyTorch was compiled without CUDA");
+
   pushCallback(
-      [config](const RecordFunction& fn) {
+      [new_config](const RecordFunction& fn) {
```

Comment:
why do you capture `new_config` in begin function but use the global config in end function?

Reply:
thanks, good catch